tests/unit/s2n_async_pkcs11_sign_test.c

```diff
@@ -0,0 +1,574 @@
+
+    struct s2n_connection *conn = info->conn;
+    struct s2n_async_pkey_op *op = info->op;
+
+    uint32_t input_len;
+    s2n_async_pkey_op_get_input_size(op, &input_len);
+    uint8_t * input = malloc(input_len);
+
+    s2n_async_pkey_op_get_input(op, input,input_len);
+
+    uint8_t * output = NULL;
+    uint32_t output_len = 0;
+
+    s2n_async_pkey_op_type type;
+    s2n_async_get_op_type(op, &type);
+
+    pthread_mutex_lock(&pkcs11_mutex);
+
+    CK_OBJECT_HANDLE handle;
+    CK_SESSION_HANDLE session;
+
+    struct s2n_cert_chain_and_key * cert_key = s2n_connection_get_selected_cert(conn);
+    char * label = s2n_cert_chain_and_key_get_ctx(cert_key);
+
+    pkcs11_setup(&session, &handle, label, strlen(label));
+    if(type == S2N_ASYNC_DECRYPT)
+    {
+        pkcs11_decrypt(session, handle, input, input_len, &output, &output_len);
+    } else {
+        pkcs11_sign(session, handle, input, input_len, &output, &output_len);
+    }
+    pthread_mutex_unlock(&pkcs11_mutex);
+
+    s2n_async_pkey_copy_output(op, output, output_len);
+    free(output);
+
+    s2n_async_pkey_op_apply(op, conn);
+    
+    free(params);
+    pthread_exit(NULL);
+}
+
+int async_pkey_callback(struct s2n_connection *conn, struct s2n_async_pkey_op *op)
+{
+    EXPECT_NOT_NULL(op);
+    pthread_t worker;
+    struct task_params * params = malloc(sizeof(struct task_params));
+    params->conn = conn; 
+    params->op = op; 
+
+    POSIX_GUARD(pthread_create(&worker, NULL, &pkey_task, params));
+
+    return S2N_SUCCESS;
+}
+
+int main(int argc, char **argv)
+{
+    BEGIN_TEST();
+    EXPECT_SUCCESS(s2n_disable_tls13());
+
+    char dhparams_pem[S2N_MAX_TEST_PEM_SIZE];
+    EXPECT_SUCCESS(s2n_read_test_pem(S2N_DEFAULT_TEST_DHPARAMS, dhparams_pem, S2N_MAX_TEST_PEM_SIZE));
+
+    struct s2n_cert_chain_and_key *rsa_chain_and_key;
+    struct s2n_cert_chain_and_key *ecdsa_chain_and_key;
+    char rsa_cert_chain_pem[S2N_MAX_TEST_PEM_SIZE];
+    char ecdsa_cert_chain_pem[S2N_MAX_TEST_PEM_SIZE];
+
+    POSIX_GUARD(s2n_read_test_pem(S2N_DEFAULT_TEST_CERT_CHAIN, rsa_cert_chain_pem, S2N_MAX_TEST_PEM_SIZE));
+    POSIX_GUARD(s2n_read_test_pem(S2N_ECDSA_P256_PKCS1_CERT_CHAIN, ecdsa_cert_chain_pem, S2N_MAX_TEST_PEM_SIZE));
+
+    POSIX_GUARD_PTR(rsa_chain_and_key = s2n_cert_chain_and_key_new());
+    POSIX_GUARD_PTR(ecdsa_chain_and_key = s2n_cert_chain_and_key_new());
+    POSIX_GUARD(s2n_cert_chain_load_pem(rsa_chain_and_key, rsa_cert_chain_pem));
+    POSIX_GUARD(s2n_cert_chain_load_pem(ecdsa_chain_and_key, ecdsa_cert_chain_pem));
+
+    POSIX_GUARD(s2n_cert_chain_and_key_set_ctx(rsa_chain_and_key, pkcs11_rsa_key_label));
+    POSIX_GUARD(s2n_cert_chain_and_key_set_ctx(ecdsa_chain_and_key, pkcs11_ec_key_label));
+
+    struct s2n_cipher_suite *test_cipher_suites[] = {
+        &s2n_rsa_with_aes_128_gcm_sha256,
+        &s2n_ecdhe_rsa_with_aes_128_gcm_sha256,
+        &s2n_ecdhe_ecdsa_with_aes_128_cbc_sha256,
+        &s2n_ecdhe_ecdsa_with_aes_256_cbc_sha384,
+    };
```

Comment:
Since this is a signing test, why are you more interested in cipher preferences than signature preferences?

If you are sure you care about cipher suites, then include a TLS1.3 one too.

Reply:
I will look into this, I derived this from the existing async_pkey test